### DISCUSSION TITLE
Cannot send commands to ESP32 after connection.

### DISCUSSION THREAD

**Discussion by shnlmn** [micropython/micropython#12631] [Libraries & Drivers]
_EDIT: 
While I wouldn't call it solved, I think my particular NodeMCU boards are the problem. I went and [bought these boards](https://www.amazon.com/gp/product/B09GK74F7N/ref=ppx_yo_dt_b_asin_title_o00_s01?ie=UTF8&psc=1), same chip but different board, and (so far) it works just fine._
____

I'm a semi-beginner with Linux and esp32. I'm working on a ESP32 WROOM and using rshell and/or mpremote on Ubuntu.

The problem: I'm finding that sending a command to the board only works some of the time, usually after I've pressed the reset button. It seems somewhat random when mpremote commands will work and when they won't. Generally, if I reset the board, wait a second, and send a command everything is fine, but the next will just hang. For example, I just did this and it just hangs until I Ctl-C.
```
➜  AP_RX git:(main) ✗ mp u0 ls
ls :
       12288 .upload.sh.swp
          14 board.py
           0 boot.py
         829 connect.sh
        1902 main copy.py
        1985 main.py
          32 wificred.py
➜  AP_RX git:(main) ✗ mp u0 ls
```
Another clue is that the script inside bilnks the board LED until it connects to Wifi, so when the board is reset the LED blinks... this (usually) means I will connect successfully. But once the command is done, the blinking stops and I cannot connect again.

Lastly, it once came out of hang on its own and it gave me the error:
```
    raise TransportError("could not enter raw repl")
mpremote.transport.TransportError: could not enter raw repl
```
I'm at a loss, perhaps I'm missing something essential about board development. Can anyone explain this?

UPDATE:
More clarity 
```
~$ mpremote --version
mpremote 1.21.0
~$ python3 --version
Python 3.10.12
~$ python3
>>> serial.VERSION
'3.5'
```
**Steps to repeat**
Connect ESP32 board to PC
Flash Micropython bootloader [v1.21.0 (2023-10-05) .bin](https://micropython.org/resources/firmware/ESP32_GENERIC-20231005-v1.21.0.bin)
In terminal:
```
~$ mpremote u0
MicroPython v1.20.0 on 2023-04-26; ESP32 module with ESP32
Type "help()" for more information.
>>> 
```
(CTRL-X)
```
~$ mpremote u0
Connected to MicroPython at /dev/ttyUSB0
Use Ctrl-] or Ctrl-x to exit this shell
```
At this point it is unresponsive to CTRL-C. CTRL-X will drop back to shell. 
This also happens with any command that connects to the board
```
~$ mpremote u0 ls
ls :
          14 board.py
           0 boot.py
         829 connect.sh
        1985 main.py
          32 wificred.py
~$ mpremote u0 ls
```
**[no response. CTRL-C to end loop]**
```
^CTraceback (most recent call last):
  File "/usr/local/bin/mpremote", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.10/dist-packages/mpremote/main.py", line 528, in main
    handler_func(state, args)
  File "/usr/local/lib/python3.10/dist-packages/mpremote/commands.py", line 110, in do_filesystem
    state.ensure_raw_repl()
  File "/usr/local/lib/python3.10/dist-packages/mpremote/main.py", line 470, in ensure_raw_repl
    self.transport.enter_raw_repl(soft_reset=soft_reset)
  File "/usr/local/lib/python3.10/dist-packages/mpremote/transport_serial.py", line 152, in enter_raw_repl
    data = self.read_until(1, b"raw REPL; CTRL-B to exit\r\n>")
  File "/usr/local/lib/python3.10/dist-packages/mpremote/transport_serial.py", line 118, in read_until
    data = self.serial.read(min_num_bytes)
  File "/home/shane/.local/lib/python3.10/site-packages/serial/serialposix.py", line 565, in read
    ready, _, _ = select.select([self.fd, self.pipe_abort_read_r], [], [], timeout.time_left())
KeyboardInterrupt
```
I recieved an ESP32-S3, which uses ACM ports. It works as expected. Only boards that show up as USB ports seem to be effected.

**Comment by davefes** [comment]
Is the ESP32 WROOM on a dev board, if so, which one?

**Reply by shnlmn** [reply]
A NodeMCU ESP32s v1.1

**Reply by davefes** [reply]
Not sure if I am looking at the correct schematic as I don't see the transistor-pair that handle the reset/boot switching.  Could you link to your schematic?

**Reply by shnlmn** [reply]
Your link below was correct.
https://wiki.geekworm.com/NodeMCU-32S

**Comment by davefes** [comment]
`main copy.py`

I would rename that `main_copy.py`

**Reply by davefes** [reply]
https://wiki.geekworm.com/NodeMCU-32S
correct?

**Reply by shnlmn** [reply]
That is the one.

**Reply by davefes** [reply]
It has the proper transistor-pair switching on it.  Looks much the same as the boards I use.  I only use rshell.

It is helpful to place this at the beginning of your program:
```py
print ('waiting 5 seconds so that you can do a CTRL-C')
time.sleep(5)
```
Gives you a chance to break-in and modify things before the program starts to run.

**Reply by davefes** [reply]
There are a couple of hardware-related differences between that board and the ones I use.  Do other people have issues with that board?

**Reply by shnlmn** [reply]
rshell gives me the same issues, except hitting reset fixes it during the hang. I'm noticing what when the command finishes, like `ls`, it (I think) sends a CTRL-C to the board. It is only after this happens that I cannot connect anymore.
For example, if I enter REPL and hit CTRL-C to exit, none of the commands will work.
I'm pretty new to microcontrollers, so I don't really understand what is happening here. Perhaps another board will work fine?

**Reply by shnlmn** [reply]
I may have been mixing up terms. I'm always connected, but sending remote commands using mpremote or rshell does not work. I can always get to the REPL.

**Comment by robert-hh** [comment]
It does not seem to be a problem with the boot switch transistors, because then there would be no communication at all.  You seem to use Thonny. Thonny does a lot of magic under the hood, which may cause boards to fail. Try a simple terminal emulator instead, like mpremote, Putty, screen, or picocom.

**Reply by shnlmn** [reply]
I've never used Thonny. I'm using Vim and everything is in the terminal. I was using Pymakr on Windows, but really wanted to take advantage of the linux support for development boards, so I bought a small computer and installed minimal Ubuntu on that. I have tried to use only the bare minimum on the linux machine.

I'm beginning to think it is the boards, though I don't really have any other of the same type to test. I have some ESP S2 Mini which don't have the same issue, but hey are ACM ports, so different protocols. 

It might be knock-off boards, I got them off Amazon and who knows anymore, but they've otherwise been performing just fine.

**Reply by robert-hh** [reply]
If there is a modem manager in your Linux system, please try to remove it. They interfere with the serial interface, even if they usually only affect the ACM devices. But removing it cannot hurt.

**Reply by robert-hh** [reply]
And please try a very simple terminal emulator like picocom.

**Reply by shnlmn** [reply]
I can start the repl through mpremote or rshell and interact with it just fine, so picocom always works. It doesn't seem to be an issue with connection, just sending commands like `ls` from rshell or mp remote.

**Reply by robert-hh** [reply]
Does the board also lock up after Ctrl-C if you use picocom?

**Reply by robert-hh** [reply]
What happens with mpremote if you specify the port, like for /dev/USB0 `mpremote u0 ls`?

**Reply by shnlmn** [reply]
Locks up in both cases. And an update: I can get to REPL, or at least get to
```
➜  ~ mp u0
Connected to MicroPython at /dev/ttyUSB0
Use Ctrl-] or Ctrl-x to exit this shell
```
but CTRL-C will not unfreeze it without hitting reset... Something about whatever the programs are sending to the board when the commands or connection is finished is locking up the board.

**Reply by robert-hh** [reply]
If you have a logic analyzer you get hook it up to the RX pin. That's the input to the board, and then you see what the board gets.

**Reply by robert-hh** [reply]
Turns out I had a board of the same type in my drawers. That works fine.

**Reply by shnlmn** [reply]
I have no logic analyzer... that's a little more advanced than I'm at this moment.

**Comment by shariltumin** [comment]
I am using LinuxMint. Just wondering what version of mpremote and python do you have?

Here are mine:
```
$ mpremote --version
mpremote 0.4.0

$ python --version
Python 3.10.2

```

I do not have the problem you have.

```
$ mpremote u0 ls
ls :
         139 boot.py
        1042 mulcast_server_mp.py
        1306 mulcast_server_mp_esp32.py
           8 tnet.sec
          21 wifi.sec
$ mpremote u0 cat boot.py
# This file is executed on every boot (including wake-boot from deepsleep)
#import esp
#esp.osdebug(None)
#import webrepl
#webrepl.start()
$ mpremote u0
Connected to MicroPython at /dev/ttyUSB0
Use Ctrl-] to exit this shell

>>> 
```

With rshell:

```
$ rshell --version
0.0.31
$ rshell -p /dev/ttyUSB0
Using buffer-size of 32
Connecting to /dev/ttyUSB0 (buffer-size 32)...
Trying to connect to REPL  connected
Retrieving sysname ... esp32
Testing if ubinascii.unhexlify exists ... Y
Retrieving root directories ... /boot.py/ /mulcast_server_mp.py/ /mulcast_server_mp_esp32.py/ /tnet.sec/ /wifi.sec/
Setting time ... Oct 09, 2023 21:06:39
Evaluating board_name ... pyboard
Retrieving time epoch ... Jan 01, 2000
Welcome to rshell. Use Control-D (or the exit command) to exit rshell.
/home/sharil/python> help

Documented commands (type help <topic>):
========================================
args    cat  connect  date  edit  filesize  help  mkdir  rm     shell
boards  cd   cp       echo  exit  filetype  ls    repl   rsync

Use Control-D (or the exit command) to exit rshell.
/home/sharil/python> repl
Entering REPL. Use Control-X to exit.
>
MicroPython v1.22.0-preview.1.g523284777.dirty on 2023-10-06; ESP32 ACT/BLE KAKI5 with ESP32
>>> 
>>>  
```

Perhaps your **pyserial** module is not working properly?

```
$ python
Python 3.10.2 (main, Feb  4 2022, 18:51:31) [GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import serial
>>> serial
<module 'serial' from '/home/sharil/opt/lib/python3.10/site-packages/serial/__init__.py'>
>>> serial.VERSION
'3.5'
>>>
```

**Reply by shnlmn** [reply]
Thanks for the question, I should have included my versions. Here are all my versions for comparison

```
➜  AP_RX git:(main) ✗ mpremote --version
mpremote 1.21.0
➜  AP_RX git:(main) ✗ python3 --version
Python 3.10.12
>>> serial.VERSION
'3.5'
```

You are on a much older version of mpremote than I am, maybe I should try a rollback.
